From aca50c2fb144dc9eba6cc1338205c68b4a38f41b Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:47 -0400
Subject: [powerpc] Disable doorbells on Power8 DD1.x

Message-id: <1410545655-205645-599-git-send-email-dzickus@redhat.com>
Patchwork-id: 94457
O-Subject: [RHEL7 PATCH 598/626] powerpc: Disable doorbells on Power8 DD1.x
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit bd6ba3518fcb2539d83163a3f486d09411bc535d
Author: Joel Stanley <joel@jms.id.au>
Date:   Fri Jul 18 11:41:37 2014 +0930

    powerpc: Disable doorbells on Power8 DD1.x

    These processors do not currently support doorbell IPIs, so remove them
    from the feature list if we are at DD 1.xx for the 0x004d part.

    This fixes a regression caused by d4e58e5928f8 (powerpc/powernv: Enable
    POWER8 doorbell IPIs). With that patch the kernel would hang at boot
    when calling smp_call_function_many, as the doorbell would not be
    received by the target CPUs:

      .smp_call_function_many+0x2bc/0x3c0 (unreliable)
      .on_each_cpu_mask+0x30/0x100
      .cpuidle_register_driver+0x158/0x1a0
      .cpuidle_register+0x2c/0x110
      .powernv_processor_idle_init+0x23c/0x2c0
      .do_one_initcall+0xd4/0x260
      .kernel_init_freeable+0x25c/0x33c
      .kernel_init+0x1c/0x120
      .ret_from_kernel_thread+0x58/0x7c

    Fixes: d4e58e5928f8 (powerpc/powernv: Enable POWER8 doorbell IPIs)
    Signed-off-by: Joel Stanley <joel@jms.id.au>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/cputable.h b/arch/powerpc/include/asm/cputable.h
index 8b27acc..1e02b2c 100644
--- a/arch/powerpc/include/asm/cputable.h
+++ b/arch/powerpc/include/asm/cputable.h
@@ -442,6 +442,7 @@ extern const char *powerpc_base_platform;
      CPU_FTR_DBELL | CPU_FTR_HAS_PPR | CPU_FTR_DAWR | \
      CPU_FTR_ARCH_207S | CPU_FTR_TM_COMP)
 #define CPU_FTRS_POWER8E (CPU_FTRS_POWER8 | CPU_FTR_PMAO_BUG)
+#define CPU_FTRS_POWER8_DD1 (CPU_FTRS_POWER8 & ~CPU_FTR_DBELL)
 #define CPU_FTRS_CELL (CPU_FTR_USE_TB | CPU_FTR_LWSYNC | \
      CPU_FTR_PPCAS_ARCH_V2 | CPU_FTR_CTRL | \
      CPU_FTR_ALTIVEC_COMP | CPU_FTR_MMCRA | CPU_FTR_SMT | \
diff --git a/arch/powerpc/kernel/cputable.c b/arch/powerpc/kernel/cputable.c
index 08e4fc8..c279a5e 100644
--- a/arch/powerpc/kernel/cputable.c
+++ b/arch/powerpc/kernel/cputable.c
@@ -527,6 +527,26 @@ static struct cpu_spec __initdata cpu_specs[] = {
   .machine_check_early = __machine_check_early_realmode_p8,
   .platform  = "power8",
  },
+ { /* Power8 DD1: Does not support doorbell IPIs */
+  .pvr_mask  = 0xffffff00,
+  .pvr_value  = 0x004d0100,
+  .cpu_name  = "POWER8 (raw)",
+  .cpu_features  = CPU_FTRS_POWER8_DD1,
+  .cpu_user_features = COMMON_USER_POWER8,
+  .cpu_user_features2 = COMMON_USER2_POWER8,
+  .mmu_features  = MMU_FTRS_POWER8,
+  .icache_bsize  = 128,
+  .dcache_bsize  = 128,
+  .num_pmcs  = 6,
+  .pmc_type  = PPC_PMC_IBM,
+  .oprofile_cpu_type = "ppc64/power8",
+  .oprofile_type  = PPC_OPROFILE_INVALID,
+  .cpu_setup  = __setup_cpu_power8,
+  .cpu_restore  = __restore_cpu_power8,
+  .flush_tlb  = __flush_tlb_power8,
+  .machine_check_early = __machine_check_early_realmode_p8,
+  .platform  = "power8",
+ },
  { /* Power8 */
   .pvr_mask  = 0xffff0000,
   .pvr_value  = 0x004d0000,
-- 
1.7.1